lib/elixir/lib/calendar/date.ex

```diff
@@ -254,6 +256,12 @@ defmodule Date do
     end
   end
 
+  def from_iso8601(<<?-, next, rest::binary>>, calendar) when next != ?- do
+    with {:ok, %{year: year} = date} <- from_iso8601(<<next>> <> rest, calendar) do
```

Comment:
I think it will be better to not assert on the `next` here, but instead assert the year is positive after parsing.

The difference is that right now we'll do an allocation (to rebuild the binary) in the expected, positive case. Matching on the value after parsing means we could do additional work if there are many leading `-`, but it is not an expected case.

Alternatively, we could extract parsing to a separate function:
```
@compile {:inline, from_iso8601_positive: 1}